hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/RegionProcedureStore.java

```diff
@@ -0,0 +1,576 @@
+      LOG.error(HBaseMarkers.FATAL, "Failed to update proc {}", proc, e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void delete(long procId) {
+    try {
+      region
+        .put(new Put(Bytes.toBytes(procId)).addColumn(FAMILY, PROC_QUALIFIER, EMPTY_BYTE_ARRAY));
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete pid={}", procId, e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void delete(Procedure<?> parentProc, long[] subProcIds) {
+    List<Mutation> mutations = new ArrayList<>(subProcIds.length + 1);
+    List<byte[]> rowsToLock = new ArrayList<>(subProcIds.length + 1);
+    try {
+      serializePut(parentProc, mutations, rowsToLock);
+      for (long subProcId : subProcIds) {
+        serializeDelete(subProcId, mutations, rowsToLock);
+      }
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete parent proc {}, sub pids={}", parentProc,
+        Arrays.toString(subProcIds), e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void delete(long[] procIds, int offset, int count) {
+    if (count == 0) {
+      return;
+    }
+    if (count == 1) {
+      delete(procIds[offset]);
+      return;
+    }
+    List<Mutation> mutations = new ArrayList<>(count);
+    List<byte[]> rowsToLock = new ArrayList<>(count);
+    for (int i = 0; i < count; i++) {
+      long procId = procIds[offset + i];
+      serializeDelete(procId, mutations, rowsToLock);
+    }
+    try {
+      region.mutateRowsWithLocks(mutations, rowsToLock, NO_NONCE, NO_NONCE);
+    } catch (IOException e) {
+      LOG.error(HBaseMarkers.FATAL, "Failed to delete pids={}", Arrays.toString(procIds), e);
+      throw new UncheckedIOException(e);
+    }
+    flusherAndCompactor.onUpdate();
+  }
+
+  @Override
+  public void cleanup() {
+    // actually delete the procedures if it is not the one with the max procedure id.
+    List<Cell> cells = new ArrayList<Cell>();
+    try (RegionScanner scanner =
+      region.getScanner(new Scan().addColumn(FAMILY, PROC_QUALIFIER).setReversed(true))) {
+      // skip the row with max procedure id
+      boolean moreRows = scanner.next(cells);
+      if (cells.isEmpty()) {
+        return;
+      }
+      cells.clear();
+      while (moreRows) {
+        moreRows = scanner.next(cells);
+        if (cells.isEmpty()) {
+          continue;
+        }
+        Cell cell = cells.get(0);
+        cells.clear();
+        if (cell.getValueLength() == 0) {
+          region.delete(new Delete(cell.getRowArray(), cell.getRowOffset(), cell.getRowLength()));
+        }
+      }
+    } catch (IOException e) {
+      LOG.warn("Failed to clean up delete procedures", e);
+    }
+  }
+}
```

Comment:
Debugging, is there means of seeing state of Procedures? Is that Master UI? Earlier ask about pretty printer.

Reply:
You can see the state on master UI.